compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -646,6 +647,18 @@ object Checking {
       val enumCls = enumCase.owner.linkedClass
       if !cls.info.parents.exists(_.typeSymbol == enumCls) then
         ctx.error(i"enum case does not extend its enum $enumCls", enumCase.sourcePos)
+
+  /** Check the inline override methods only use inline parameteres if they override an inline parameter. */
+  def checkInlineOverrideParameters(sym: Symbol)(using Context): Unit =
+    val params = sym.paramSymss.flatten
+    if params.exists(_.is(Inline)) then
+      for
+        sym2 <- sym.allOverriddenSymbols
+        (p1, p2) <- params.lazyZip(sym2.paramSymss.flatten)
+        if p1.is(Inline) && !p2.is(Inline)
```

Comment:
What about the reverse? Is it ok for a non-inline parameter to override an inline parameter?

Reply:
It's debatable whether it's OK or desirable. But at least it is *sound*, because, as @nicolasstucki observed, you can always emulate the semantics of a non-`inline` param with an `inline` param that you immediately assign to a `val` in the body of the method.

However, personally I think we should not allow it either. In Scala, typically overrides must have exactly the same parameters. I think we should also apply that to the quality of being `inline`. And *if* someone really needs that capability, well they can always do the emulation themselves.

Reply:
Well, actually I think they're unsound. If we have a method with two parameters, the first one being either `inline`+`val` or by-value, we can end up with different evaluation order for the actual arguments:
```scala
def foo(inline x: Int, y: Int): Int = {
  val x2 = x
  x2 * x2 + y
}

foo({ println(1); 3 }, { println(2); 4 })
// prints 2
// prints 1
// returns 13
```
but
```scala
def foo(x: Int, y: Int): Int = {
  x * x + y
}

foo({ println(1); 3 }, { println(2); 4 })
// prints 1
// prints 2
// returns 13
```
So overriding an inline parameter with a by-value parameter is unsound.

Reply:
I already added a commit disallowing them.